modules/core/src/main/scala/gem/Location.scala

```diff
@@ -10,46 +11,171 @@ import scalaz.Ordering.{EQ, GT, LT}
+sealed trait Location {
 
-  // Use the sketchy sealed abstract case class technique to control the
-  // NonEmptyList values that are passed to the constructor.  We want to
-  // ensure that trailing Int.MinValue is always trimmed so that EQ and ==
-  // agree.
+  // These functions aren't of any use to clients.  Instead they are involved
+  // in the cacluation of Locations that fall between two other locations.
 
-  def toList: List[Int] =
-    toIList.toList
+  /** Infinite Stream of position elements corresponding to this Location. */
+  protected def positions: Stream[Int]
 
-  override def toString: String =
-    toList.mkString("{", ",", "}")
+  /** Minimum size required to obtain all the fixed elements in this Location,
+    * if any.  Beginning and End Locations have no fixed elements.
+    */
+  protected def minPrefixLength: Int
 }
 
 object Location {
-  private def trim(is: Seq[Int]): IList[Int] =
-    trim(IList(is: _*))
 
-  // Remove trailing Int.MinValue so that == agrees with EQ.
-  private def trim(is: IList[Int]): IList[Int] =
-    is.dropRightWhile(_ == Int.MinValue)
+  /** A marker for the first Location.  No other Location comes before the
+    * Beginning.  Beginning should not be used to order elements unless you
+    * are sure no elements will ever need to be inserted before Beginning.  Use
+    * Beginning to calculate Locations that fall before the first existing
+    * Location (if any).
+    */
+  case object Beginning extends Location {
+    protected def positions: Stream[Int] =
+      Stream.continually(Int.MinValue)
+
+    protected def minPrefixLength: Int =
+      0
+
+    override def toString: String =
+      "{α}"
+  }
+
+  /** A Location that falls somewhere between the Beginning and End.
+    */
+  sealed abstract case class Middle(posList: NonEmptyList[Int]) extends Location {
+    def positions: Stream[Int] =
+      posList.toStream #::: Stream.continually(Int.MinValue)
+
+    def toIList: IList[Int] =
+      posList.list
+
+    def toList: List[Int] =
+      toIList.toList
+
+    protected def minPrefixLength: Int =
+      posList.size
+
+    override def toString =
+      toList.mkString("{", ",", "}")
+  }
+
+  /** A marker for the last Location.  No other Location comes after the End.
+    * End should not be used to order elements unless you are sure no elements
+    * will ever need to be inserted after End.  Use End to calculate new
+    * Locations that fall after the last existing Location (if any).
+    */
+  case object End extends Location {
+    def positions: Stream[Int] =
+      Stream.continually(Int.MaxValue)
+
+    protected def minPrefixLength: Int =
+      0
+
+    override def toString: String =
+      "{ω}"
+  }
+
+  // Constructors
+
+  private implicit object FoldableSeq extends Foldable[Seq] with Foldable.FromFoldr[Seq] {
+    def foldRight[A, B](fa: Seq[A], z: => B)(f: (A, => B) => B): B =
+      fa.foldRight(z) { (a, b) => f(a, b) }
+  }
```

Comment:
Yes, scalaz has no instances for `Seq` because it's mutable. It's less of an issue for `Foldable` because it's hard to identify laws, but still doesn't really fit with the way scalaz thinks about things.

In 2.13 the default `Seq` is going to become an alias for `immutable.Seq` so with some luck this will be fixed. Anyway not a big deal if it's local.